controlplane/kubeadm/controllers/controller.go

```diff
@@ -282,17 +282,27 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 		return ctrl.Result{}, nil
 	}
 
-	controlPlane := internal.NewControlPlane(cluster, kcp, ownedMachines)
+	infraObjs, err := r.getInfraObjects(ctx, ownedMachines)
+	if err != nil {
+		logger.Error(err, "failed to retrieve infra objects")
+		return ctrl.Result{}, err
+	}
+	machineConfigs, err := r.getMachineConfigs(ctx, ownedMachines)
+	if err != nil {
+		logger.Error(err, "failed to retrieve bootstrap machine configs")
+		return ctrl.Result{}, err
+	}
+	controlPlane := internal.NewControlPlane(cluster, kcp, ownedMachines, machineConfigs, infraObjs)
 
 	// Aggregate the operational state of all the machines; while aggregating we are adding the
 	// source ref (reason@machine/name) so the problem can be easily tracked down to its source machine.
 	conditions.SetAggregate(controlPlane.KCP, controlplanev1.MachinesReadyCondition, ownedMachines.ConditionGetters(), conditions.AddSourceRef())
 
 	// Control plane machines rollout due to configuration changes (e.g. upgrades) takes precedence over other operations.
-	needRollout := r.machinesNeedingRollout(ctx, controlPlane)
+	needRollout := controlPlane.MachinesNeedingRollout()
```

Comment:
Yeah I tried to push back on it moving back to the reconciler but I had a hard time explaining the rationale. There was a bug with `time.Now` changing that we fixed by moving it back and threading the results of `needsRollout` through to the controller rather than calling the method again. I didn't change the passing of `needsRollout` back here.

Reply:
If you're actually serious that we should have a comment about this I'm not sure how to word that. Saying "please don't move me back to the reconciler" sounds weird :smil

Reply:
It'd be great to have at least some wording on ControlPlane on why to prefer this struct vs having everything on the reconciler

Reply:
Added it in a4538144c03bba6d01e103e285772628ac6fc108